### PR TITLE
libdistance 0.2.2 (new formula)

### PR DESCRIPTION
**Pull request by trombonehero**
Create a new formula for Jose Nazario's library for calculating edit
distances between strings.

### CODE REVIEW

**Comment by mistydemeo** [comment]
Is it feasible to write a test for this?

File: Library/Formula/libdistance.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class Libdistance < Formula
+  homepage "http://monkey.org/~jose/software/libdistance/"
+  url "http://monkey.org/~jose/software/libdistance/libdistance-0.2.2.tar.gz"
+  sha1 "d516e5f24ae55bcbbf93a4d14c6a53914e5b3a0a"
+
+  def install
+    # The build system is incredibly simple.
+    system "make"
+    include.install "distance.h"
```
Comment (mikemcquaid) [inline Library/Formula/libdistance.rb]: Is there no `makeinstall`?

File: Library/Formula/libdistance.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    lib.install "libdistance.a"
+  end
+end
```
Comment (mikemcquaid) [inline Library/Formula/libdistance.rb]: Can this test be added to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

File: Library/Formula/libdistance.rb (hunk 1 above)
Comment (trombonehero) [inline Library/Formula/libdistance.rb]: Not in the GNU makefile, nope.

The BSD makefile does have an `install` target via `bsd.lib.mk`, but that Makefile builds tcl bindings via swig, which is overkill when all I want is the simple C library.

Alternatively, can I create a --with-swig option and then choose between the build systems appropriately? (sorry for an elementary question, this is my first Homebrew formula)

File: Library/Formula/libdistance.rb (hunk 2 above)
Comment (trombonehero) [inline Library/Formula/libdistance.rb]: libdistance actually ships with a test suite, but it seems that the original sources are gone by the time that `Libdistance.test` is run. If we build the test program, can it be installed to the Cellar but not /usr/local/bin?

**Comment by trombonehero** [comment]
Hi @mistydemeo, @mikemcquaid,

I've modified things so that the libdistance test suite will be built and copied to the Cellar. The GNU makefile included with the project doesn't include `install` targets, etc., so I've used some FileUtils functions instead. One question: can I pass `bin.install` a parameter to mean "put this file in the Cellar but don't symlink it to /usr/local/bin"?

**Comment by mikemcquaid** [comment]
> One question: can I pass bin.install a parameter to mean "put this file in the Cellar but don't symlink it to /usr/local/bin"?

No.

Given the installation instructions (http://monkey.org/~jose/software/libdistance/#installation) say to `make install` I'd rather if we did that (even if it requires some patches which we submit upstream)

File: Library/Formula/libdistance.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+class Libdistance < Formula
+  homepage "http://monkey.org/~jose/software/libdistance/"
+  url "http://monkey.org/~jose/software/libdistance/libdistance-0.2.2.tar.gz"
+  sha1 "d516e5f24ae55bcbbf93a4d14c6a53914e5b3a0a"
+
+  def install
+    # The build system is incredibly simple.
```
Comment (bfontaine) [inline Library/Formula/libdistance.rb]: You can remove this comment.

File: Library/Formula/libdistance.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    system "make"
+    system "make -C test"
```
Comment (bfontaine) [inline Library/Formula/libdistance.rb]: Please use `system "make", "-C", "test"`.

**Comment by bfontaine** [comment]
@BrewTestBot test this please.

**Comment by mikemcquaid** [comment]
Some audit warnings need fixed: http://bot.brew.sh/job/Homebrew%20Pull%20Requests/27020/version=yosemite/testReport/junit/brew-test-bot/yosemite/audit_libdistance___strict/

File: Library/Formula/libdistance.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+    include.install "distance.h"
```
Comment (mikemcquaid) [inline Library/Formula/libdistance.rb]: Is there no `make install` we can use?

File: Library/Formula/libdistance.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    lib.install "libdistance.a"
+
+    mkdir "#{prefix}/test"
+    cp "test/test", "#{prefix}/test/"
```
Comment (mikemcquaid) [inline Library/Formula/libdistance.rb]: use `share.install` instead of `cp`.

**Comment by DomT4** [comment]
Any update to Mike and Baptiste's comments here?

File: Library/Formula/libdistance.rb (hunk 6 above)
Comment (trombonehero) [inline Library/Formula/libdistance.rb]: Hmmm, actually, changing `cp` to `share.install` broke the build in an unexpected way:

```
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:519:in `rename'
Errno::ENOTDIR: Not a directory - (/usr/local/Cellar/libdistance/0.2.2/test/, /usr/local/Cellar/libdistance/0.2.2/share/test)
```

Any ideas what this is all about?

**Comment by trombonehero** [comment]
@mikemcquaid :

The entire content of the GNUmakefile is:
```make
# $Id: GNUmakefile,v 1.5 2004/11/29 21:48:08 jose Exp $
#
# GNU Makefile, should work on mingw and other GNU make systems.

SRCS=	levenshtein.c hamming.c bloom.c needleman_wunsch.c jaccard.c \
	minkowski.c damerau.c
OBJS=	levenshtein.o hamming.o bloom.o needleman_wunsch.o jaccard.o \
	minkowski.o damerau.o

libdistance.a: ${SRCS}
	gcc -g -c -I. levenshtein.c
	gcc -g -c hamming.c
	gcc -g -c bloom.c
	gcc -g -c needleman_wunsch.c
	gcc -g -c jaccard.c
	gcc -g -c minkowski.c
	gcc -g -c damerau.c
	ar cq libdistance.a ${OBJS}
	ranlib libdistance.a

clean:
	rm -f libdistance.a libdistances.so ${OBJS} *.core
```

There is an `install` target in the BSD makefile, but that makefile also adds a lot of complexity around swig and other stuff that users of C libraries don't use. That, plus the fact that we only need two files (three if you count the test suite) is why I used `include.install` and `lib.install` rather than `make install`.

I could switch to the BSD makefile, but that would add a lot more complexity and bits.

**Comment by trombonehero** [comment]
I've just pushed a couple of commits to appease `brew audit`, but I don't know enough about Homebrew to fix the "Use new-style test definitions (test do)" warning.

File: Library/Formula/libdistance.rb (hunk 6 above)
Comment (DomT4) [inline Library/Formula/libdistance.rb]: How did you write it?

File: Library/Formula/libdistance.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    system "make", "-C", "test"
+
+    include.install "distance.h"
+    lib.install "libdistance.a"
+
+    mkdir "#{prefix}/test"
+    cp "test/test", "#{prefix}/test/"
+  end
+
+  def test
```
Comment (DomT4) [inline Library/Formula/libdistance.rb]: The `use new style` warnings means you should turn `def test` into `test do`.

**Comment by DomT4** [comment]
Still hitting some failures here:

```
==> audit problems
libdistance:
 * `desc`(line 3) should be put before `homepage`(line 2)
 * Description is too long. "name: desc" should be less than 80 characters (currently 349).
```

**Comment by bfontaine** [comment]
Any update here?

**Comment by DomT4** [comment]
Presuming dead from lack of response. Please feel free to fix-up various comments and resubmit the PR if you're still interested, but closing here for now. Thanks for the submission!

**Pull request closed without merging**